### DISCUSSION TITLE
Clustering in sktime.

### DISCUSSION THREAD

**Discussion by hstarmans** [alan-turing-institute/sktime#6421] [General]
I want to construct the following pipeline in sktime
```python
pipe = make_pipeline(StandardScaler(), PCATransformer(2), TimeSeriesKMeans(2, metric='euclidean'))
```
I want it to behave exactly the same as my pipeline in sklearn
```python
pipe = make_pipeline(StandardScaler(), PCA(2), KMeans(2))
```
Currently, sktime doesn't work very well and sklearn works like a breeze but I can't include operations like detrending etc.
I provide the following toy dataset. I have Machines, which generate features on each time step.
Basically, I am interested in anomalies and want to cluster events in for example 3 groups. This to help the business in focusing on the right problems.
A set is as follows
```python
import numpy as np
import pandas as pd
samples = 20
features = 3
base_dct = {"Time": range(samples), "Machine": [1]*samples}
for feat in range(features):
    base_dct[f'feat{feat}'] = np.random.randint(np.random.randint(3,10,size=1), size=samples)

X = pd.DataFrame(base_dct).set_index(['Machine', 'Time'])
X.head()
```
<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th></th>
      <th>feat0</th>
      <th>feat1</th>
      <th>feat2</th>
    </tr>
    <tr>
      <th>Machine</th>
      <th>Time</th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th rowspan="5" valign="top">1</th>
      <th>0</th>
      <td>1</td>
      <td>0</td>
      <td>2</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0</td>
      <td>4</td>
      <td>3</td>
    </tr>
    <tr>
      <th>2</th>
      <td>2</td>
      <td>3</td>
      <td>3</td>
    </tr>
    <tr>
      <th>3</th>
      <td>4</td>
      <td>0</td>
      <td>2</td>
    </tr>
    <tr>
      <th>4</th>
      <td>3</td>
      <td>3</td>
      <td>1</td>
    </tr>
  </tbody>
</table>
</div>
My sklearn pipe gives the result

```python
from sklearn.pipeline import make_pipeline
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
import seaborn as sns
pipe = make_pipeline(StandardScaler(), PCA(2), KMeans(2))
pipe.fit(X)
sns.pairplot(X.assign(Clusters = pipe.named_steps['kmeans'].labels_), hue='Clusters', corner=True)
```

![image](https://github.com/sktime/sktime/assets/6517246/1bcb3fd1-d2f5-4bd5-92cd-f8ab9a6f2a01)

Now in sktime
```python
from sktime.pipeline import make_pipeline
from sktime.clustering.k_means import TimeSeriesKMeans
from sktime.transformations.panel.pca import PCATransformer
from sktime.pipeline import sklearn_to_sktime
from sklearn.preprocessing import StandardScaler
from sktime.clustering.compose import ClustererPipeline
pipe = make_pipeline(StandardScaler(), PCATransformer(2), TimeSeriesKMeans(2, metric='euclidean'))
```

The first step goes fine
```python
StandardScaler().fit_transform(X).std(axis=0)
```
Hereafter it stops working. The pipeline does accept data when i do the following.
But it is not what I want
```python
X_trafo = X.reset_index(level=0, drop=True).rename(columns=dict(map(reversed, enumerate(X.columns)))).stack().to_frame().groupby(level=[0,1]).first()
pipe.fit(X_trafo)
```
Another solution I see is 
```python
pipe = make_pipeline(StandardScaler(), sklearn_to_sktime(PCA(2)), sklearn_to_sktime(KMeans(2)))
```
This works well until the PCA but fails at the kmeans as this no longer accepts 2 and raises the error.
```python
ValueError: n_samples=1 should be >= n_clusters=2.
```

**Comment by fkiraly** [comment]
The way you set it up, you are asking `sktime` to cluster entire *machines* and the behaviour is as expected.

Since there is a single machine, and you are asking `sktime` to sort it into two clusters, it recognizes this is impossible, and raises the error message.

It should work as you intend if you have 2 or more machines to cluster.

Though I'm not sure if that's what you want - perhaps the task you are interested in is not time series clustering? Do you want to apply the clustering per time index, perhaps?

That would be unsupervised annotation with categories, or segmentation.

That module exists, but it is experimental. @Alex-JG3 is currently reworking interfaces, and some summer mentees are working on it.
If this is indeed your task, how about we connect and see which algorithms you would like implemented?

**Comment by hstarmans** [comment]
@fkiraly update from my side on outlier detection in sktime
   my current pipeline consists out of a; Bandpass Filter, Kalman Filter, Differencer and the k-nearest neighbors for the outlier detection
   PyODAnnotator relies on PyOD and is not a proper time series version such as [TODS](https://github.com/datamllab/tods). It clusters per time index. I establish a view "across" time via the differencer.  Aeon has the hidalgo segmenter which is more like a clusterer across time,see https://www.nature.com/articles/s41598-020-72222-0 and  https://www.aeon-toolkit.org/en/latest/api_reference/auto_generated/aeon.segmentation.HidalgoSegmenter.html  . In the end this is not what i want, i wanted outliers.

PS.
Kalman filter is not really needed, you don't need to set variables and can it tune itself (see [basic usage](https://pykalman.github.io/#basic-usage) docs)

**Reply by fkiraly** [reply]
> Aeon has the hidalgo segmenter 

That's a copy of the `sktime` hidalgo time series segmentor (just like most of aeon is a fork/copy) - they just renamed the `annotation` module to `segmentation` - imo incorrectly, since that also includes change point detection, etc.

> my current pipeline consists out of a; Bandpass Filter, Kalman Filter, Differencer and the k-nearest neighbors for the outlier detection
PyODAnnotator relies on PyOD

Hm, interesting. We're currently reworking the time series annotation module - if you would like to contribute, we have some outlier detection algorithms we'd like to create or interface.
